session/portforward.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package session
+
+import (
+	"context"
+	"io"
+	"net"
+	"strconv"
+
+	"github.com/mendersoftware/go-lib-micro/ws"
+	wspf "github.com/mendersoftware/go-lib-micro/ws/portforward"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	protocolTCP = "tcp"
+	protocolUDP = "udp"
+)
+
+const (
+	readBuffLength = 4096
+)
+
+var (
+	errPortForwardInvalidMessage     = errors.New("invalid port-forward message: missing connection_id, remort_port or protocol")
+	errPortForwardUnkonwnMessageType = errors.New("unknown message type")
+	errPortForwardUnkonwnConnection  = errors.New("unknown connection")
+)
+
+var portForwarders = make(map[string]*MenderPortForwarder)
+
+type MenderPortForwarder struct {
+	ConnectionID   string
+	SessionID      string
+	ResponseWriter ResponseWriter
+	conn           net.Conn
+	ctx            context.Context
+	ctxCancel      context.CancelFunc
+}
+
+func (f *MenderPortForwarder) Connect(protocol string, host string, portNumber int) error {
+	log.Debugf("port-forward[%s/%s] connect: %s/%s:%d", f.SessionID, f.ConnectionID, protocol, host, portNumber)
+
+	if protocol == protocolTCP || protocol == protocolUDP {
+		conn, err := net.Dial(protocol, host+":"+strconv.Itoa(portNumber))
+		if err != nil {
+			return err
+		}
+		f.conn = conn
+	} else {
+		return errors.New("unknown protocol: " + protocol)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	f.ctx = ctx
+	f.ctxCancel = cancelFunc
+
+	go f.Read()
+
+	return nil
+}
+
+func (f *MenderPortForwarder) Close() error {
+	log.Debugf("port-forward[%s/%s] close", f.SessionID, f.ConnectionID)
+	f.ctxCancel()
```

Comment:
maybe I'm missing it, but I cannot see that the entries in the global `portForwarders` belonging to this session instance gets cleaned up.

Reply:
there are `delete(portForwarders, key)` on connection closing (or errors). Do you mean something else?

Reply:
`Close` is called whenever the session is closed - whether it's due to a fatal error or regular close. So if the connection times out, mender-connect is keeping the map entry in memory until the application is closed.
The easiest way (I can think of) is to let the garbage collector do the work by putting the global variable inside the `MenderPortForwarder` struct.

Reply:
The Close method is now responsible for removing the port forwarder from the global variable, does it make sense?

Reply:
The only issue is that you can't send messages in the `Close` handler, as it's called after the Session close condition is met. It should only deallocate resources (sorry for not making it clear earlier).

Reply:
I believe you are confusing my MenderPortForwarder (which is an internal struct to hold a single port-forwarding session) with the Session handler. Can it be?

Reply:
Yeah, sorry... My bad, I thought that `MenderPortForwarder` was implementing the `SessionHandler` interface, but I see now that you are using the `HandlerFunc` wrapper, so `Close` is indeed not the `Close` from the `SessionHandler` interface. I'll try to be more thorough next time.